File: app/src/main/java/practicaltask/riseapps/com/practicaltask/ui/regionList/RegionListPresenter.java
package practicaltask.riseapps.com.practicaltask.ui.regionList;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import practicaltask.riseapps.com.practicaltask.data.enums.Region;
import practicaltask.riseapps.com.practicaltask.storage.preferences.Preferences;
import practicaltask.riseapps.com.practicaltask.ui.base.BasePresenter;

public class RegionListPresenter extends BasePresenter<RegionListView> {

    @Inject
    protected Preferences preferences;

    @Inject
    RegionListPresenter() {
        //For DI
    }

    @Override
    protected void onViewAttached() {
        RegionListView view = getView();
        if (view != null) {
            view.showRegionList(sortWithLastOpened(new ArrayList<Region>(Arrays.asList(Region.values()))));
        }
    }

    private List<Region> sortWithLastOpened(List<Region> regions) {
        Region lastOpenedRegion = preferences.getLastOpenedRegion();
        if (lastOpenedRegion != null && regions.contains(lastOpenedRegion)) {
            regions.remove(lastOpenedRegion);
            regions.add(0, lastOpenedRegion);
        }
        return regions;
    }

    public void openRegion(Region region) {
        saveLastOpenedRegion(region);
        RegionListView view = getView();
        if (view != null) {
            view.openCountryListOfRegion(region);
        }
    }

    private void saveLastOpenedRegion(Region region) {
        preferences.saveLastOpenedRegion(region);
    }
}

File: app/src/main/java/practicaltask/riseapps/com/practicaltask/ui/regionList/RegionListPresenter.kt
package practicaltask.riseapps.com.practicaltask.ui.regionList

import practicaltask.riseapps.com.practicaltask.data.enums.Region
import practicaltask.riseapps.com.practicaltask.storage.preferences.Preferences
import practicaltask.riseapps.com.practicaltask.ui.base.BasePresenter
import javax.inject.Inject

class RegionListPresenter @Inject internal constructor() : BasePresenter<RegionListView>() {

    @Inject
    lateinit var preferences: Preferences

    override fun onViewAttached() {
        view?.showRegionList(sortWithLastOpened(Region.values().toMutableList()))
    }

    private fun sortWithLastOpened(regions: MutableList<Region>): List<Region> {
        val lastOpenedRegion = preferences.lastOpenedRegion
        if (lastOpenedRegion != null && regions.contains(lastOpenedRegion)) {
            regions.remove(lastOpenedRegion)
            regions.add(0, lastOpenedRegion)
        }
        return regions
    }

    fun openRegion(region: Region) {
        saveLastOpenedRegion(region)
        view?.openCountryListOfRegion(region)
    }

    private fun saveLastOpenedRegion(region: Region) {
        preferences.saveLastOpenedRegion(region)
    }
}
